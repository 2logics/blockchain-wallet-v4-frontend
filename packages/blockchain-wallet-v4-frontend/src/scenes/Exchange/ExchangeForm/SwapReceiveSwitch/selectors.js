import { createSelector } from 'reselect'

import { model } from 'data'

const { sourceActive, targetActive } = model.rates

export const getData = createSelector(
  (state, { fix }) => fix,
  fix => ({
    sourceActive: sourceActive(fix),
    targetActive: targetActive(fix)
  })
)
