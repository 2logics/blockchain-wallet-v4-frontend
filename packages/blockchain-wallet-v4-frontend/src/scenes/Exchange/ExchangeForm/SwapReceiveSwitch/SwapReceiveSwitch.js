import React from 'react'
import { connect } from 'react-redux'
import styled from 'styled-components'
import { FormattedMessage } from 'react-intl'

import { getData } from './selectors'

import { Cell, Row } from '../Layout'
import { Text } from 'blockchain-info-components'

const SwapReceiveRow = styled(Row)`
  padding-bottom: 0;
`
const SwapReceiveGap = styled(Cell)`
  min-width: 50px;
`
const ActiveCurrencyButton = styled.div`
  cursor: pointer;
  height: 11px;
  width: 11px;
  background-color: ${props => props.checked && props.theme[props.coin]};
  border-radius: 8px;
  margin-right: 8px;
  border: 1px solid
    ${props =>
      props.checked ? props.theme[props.coin] : props.theme['gray-4']};
`
const ClickableText = styled(Text)`
  cursor: pointer;
`

export const SwapReceiveSwitch = ({
  sourceCoin,
  targetCoin,
  sourceActive,
  targetActive,
  swapFix
}) => (
  <SwapReceiveRow>
    <Cell>
      <ActiveCurrencyButton
        data-e2e='exchangeExchangeRadioButton'
        onClick={() => {
          if (!sourceActive) swapFix()
        }}
        checked={sourceActive}
        coin={sourceCoin.toLowerCase()}
      />
      <ClickableText
        data-e2e='exchangeExchangeRadioText'
        onClick={() => {
          if (!sourceActive) swapFix()
        }}
        size='14px'
        weight={400}
      >
        <FormattedMessage
          id='scenes.exchange.exchangeform.swap'
          defaultMessage='Swap'
        />
      </ClickableText>
    </Cell>
    <SwapReceiveGap size='small' />
    <Cell>
      {
        <ActiveCurrencyButton
          data-e2e='exchangeReceiveRadioButton'
          onClick={() => {
            if (!targetActive) swapFix()
          }}
          checked={targetActive}
          coin={targetCoin.toLowerCase()}
        />
      }
      <ClickableText
        data-e2e='exchangeReceiveRadioText'
        onClick={() => {
          if (!targetActive) swapFix()
        }}
        size='14px'
        weight={400}
      >
        <FormattedMessage
          id='scenes.exchange.exchangeform.to'
          defaultMessage='Receive'
        />
      </ClickableText>
    </Cell>
  </SwapReceiveRow>
)

export default connect(getData)(SwapReceiveSwitch)
