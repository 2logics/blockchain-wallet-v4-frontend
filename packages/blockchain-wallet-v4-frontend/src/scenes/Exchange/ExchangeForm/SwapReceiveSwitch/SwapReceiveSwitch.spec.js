import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import { SwapReceiveSwitch } from './SwapReceiveSwitch'

const swapFix = jest.fn()

describe('SwapReceiveSwitch', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })
  it('should render correctly', () => {
    const component = shallow(
      <SwapReceiveSwitch
        sourceCoin='BTC'
        targetCoin='XLM'
        sourceActive={true}
        targetActive={false}
        swapFix={swapFix}
      />
    )

    const tree = toJson(component)
    expect(tree).toMatchSnapshot()
  })

  it('should call swap fix upon swap button or text if source is not active', () => {
    const component = shallow(
      <SwapReceiveSwitch
        sourceCoin='BTC'
        targetCoin='XLM'
        sourceActive={false}
        targetActive={true}
        swapFix={swapFix}
      />
    )

    const button = component.find('[data-e2e="exchangeExchangeRadioButton"]')
    const text = component.find('[data-e2e="exchangeExchangeRadioText"]')

    button.simulate('click')
    expect(swapFix).toHaveBeenCalledTimes(1)

    text.simulate('click')
    expect(swapFix).toHaveBeenCalledTimes(2)
  })

  it('should not call swap fix upon swap button or text if source is active', () => {
    const component = shallow(
      <SwapReceiveSwitch
        sourceCoin='BTC'
        targetCoin='XLM'
        sourceActive={true}
        targetActive={false}
        swapFix={swapFix}
      />
    )

    const button = component.find('[data-e2e="exchangeExchangeRadioButton"]')
    const text = component.find('[data-e2e="exchangeExchangeRadioText"]')

    button.simulate('click')
    expect(swapFix).toHaveBeenCalledTimes(0)

    text.simulate('click')
    expect(swapFix).toHaveBeenCalledTimes(0)
  })

  it('should call swap fix upon receive button or text if target is not active', () => {
    const component = shallow(
      <SwapReceiveSwitch
        sourceCoin='BTC'
        targetCoin='XLM'
        sourceActive={true}
        targetActive={false}
        swapFix={swapFix}
      />
    )

    const button = component.find('[data-e2e="exchangeReceiveRadioButton"]')
    const text = component.find('[data-e2e="exchangeReceiveRadioText"]')

    button.simulate('click')
    expect(swapFix).toHaveBeenCalledTimes(1)

    text.simulate('click')
    expect(swapFix).toHaveBeenCalledTimes(2)
  })

  it('should not call swap fix upon receive button or text if target is active', () => {
    const component = shallow(
      <SwapReceiveSwitch
        sourceCoin='BTC'
        targetCoin='XLM'
        sourceActive={false}
        targetActive={true}
        swapFix={swapFix}
      />
    )

    const button = component.find('[data-e2e="exchangeReceiveRadioButton"]')
    const text = component.find('[data-e2e="exchangeReceiveRadioText"]')

    button.simulate('click')
    expect(swapFix).toHaveBeenCalledTimes(0)

    text.simulate('click')
    expect(swapFix).toHaveBeenCalledTimes(0)
  })
})
