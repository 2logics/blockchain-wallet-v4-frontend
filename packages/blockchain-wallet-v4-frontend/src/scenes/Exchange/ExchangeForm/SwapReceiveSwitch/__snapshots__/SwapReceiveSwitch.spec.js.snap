// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwapReceiveSwitch should render correctly 1`] = `
<SwapReceiveSwitch__SwapReceiveRow>
  <Layout__Cell>
    <SwapReceiveSwitch__ActiveCurrencyButton
      checked={true}
      coin="btc"
      data-e2e="exchangeExchangeRadioButton"
      onClick={[Function]}
    />
    <SwapReceiveSwitch__ClickableText
      data-e2e="exchangeExchangeRadioText"
      onClick={[Function]}
      size="14px"
      weight={400}
    >
      <FormattedMessage
        defaultMessage="Swap"
        id="scenes.exchange.exchangeform.swap"
        values={Object {}}
      />
    </SwapReceiveSwitch__ClickableText>
  </Layout__Cell>
  <SwapReceiveSwitch__SwapReceiveGap
    size="small"
  />
  <Layout__Cell>
    <SwapReceiveSwitch__ActiveCurrencyButton
      checked={false}
      coin="xlm"
      data-e2e="exchangeReceiveRadioButton"
      onClick={[Function]}
    />
    <SwapReceiveSwitch__ClickableText
      data-e2e="exchangeReceiveRadioText"
      onClick={[Function]}
      size="14px"
      weight={400}
    >
      <FormattedMessage
        defaultMessage="Receive"
        id="scenes.exchange.exchangeform.to"
        values={Object {}}
      />
    </SwapReceiveSwitch__ClickableText>
  </Layout__Cell>
</SwapReceiveSwitch__SwapReceiveRow>
`;
